### ISSUE TITLE
[EX-33] dtask - a directory-based mode for developers and sysadmins

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#30]
_David Patrick on 2011-01-04T15:34:26Z says:_ 

 Not to be confused with taskd, dtask (or devtask) would be an iteration of taskwarrior that worked in a very directory-specific way, starting with pwd and walking all the way down the directory branches, collecting pending.data files. Unless directed otherwise, newly created tasks would be created in a pending.data file in the pwd, newly created, if necessary.

The mode is very much a tip-of-the-hat towards devtodo (http://swapoff.org/DevTodo) yet another task manager, but more targetted towards sysadmins and software developers. The directory-centric behavior allows task to be directly connected, located amongst, the source-files and settings. If your work is around these things, it can help you focus on specific parts of it merely by *cd* ing to the directory in question before issuing the query. 

I would suggest a new executable name (while invoking the same code) as the behaviors and work modes of task vs. dtask would be quite different.

(not my idea, or even new, just remembered hearing it in discussion)

NOTE: this function now available as an External Script at http://taskwarrior.org/projects/taskwarrior/wiki/ExternalScripts#local-tasksh and so status:closed

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-01-04T15:34:26Z
Modified: 2014-02-09T02:43:23Z
```

**Comment by taskwarrior** [comment]
_David Patrick on 2011-01-12T12:37:41Z says:_ 

 Aleksey Zapparov wrote:
> I don't think that this should be a core feature. 

well the way you describe it, there's no need, is there ? wow!

one caveat; for this mode to be actually useful, I think it should _aggregate_ $PWD/.task and _every pending.data file below it in the directory tree_ , otherwise it's a game of hide-the-task and you might never stumble on that one again. 
Having the query based on the branch, not just the directory, would be much more useful.
I don't know if this demands a "synthetic" pending.data file be generated, or what.

**Comment by taskwarrior** [comment]
_Aleksey Zapparov on 2011-01-29T16:23:21Z says:_ 

 Well, in this case the proposed script can be modified a little bit to allow you achieve such particular need (but to be honest, I don't see real need in it):

<pre>
1. !/bin/sh


TASKS_DIR_NAME=.task
TASKS_DIR=$PWD/$TASKS_DIR_NAME


merge_descendants()
{
    for child in `find $PWD -type d -name $TASKS_DIR_NAME | grep -v $TASKS_DIR`
    do
        echo "*** Found descendant task manager in:"
        echo "    $child"
        echo -n "    Do you want to merge it into current [y/N]: "

        read confirmation
        if [ "xy" = "x$confirmation" ] || [ "xY" = "x$confirmation" ]
        then
            for name in completed pending undo
            do
                data_file=$name.data
                if [ -e $child/$data_file ]
                then
                    cat $child/$data_file >> $TASKS_DIR/$data_file
                    rm $child/$data_file
                fi
            done
            rm -rf $child
        fi
    done
}


if [ ! -d $TASKS_DIR ]; then
1.  this hack is needed in order to tell task to use $PWD/.tasks dir as
1.  default directory, so it will put it into per-dir config. Otherwise
1.  it will use $HOME/.task
  mkdir $TASKS_DIR || exit 1
  merge_descendants
  task rc:$TASKS_DIR/rc rc.data.location:$TASKS_DIR $*
else
  merge_descendants
  task rc:$TASKS_DIR/rc $*
fi
</pre>


With this script there are few things to notice. First of all, probably it's better to ask (and show) diff of configurations, and if needed (and user wants) merge them too. Right now descendant config is simply removed. Also, +(at)for child in `...`@+ works good ONLY if there will be no spaces in the paths, otherwise it will skip some of the paths. Of course it would be better to replace:

<pre>for child in `find $PWD -type d -name $TASKS_DIR_NAME | grep -v $TASKS_DIR`</pre>

with

<pre>find $PWD -type d -name $TASKS_DIR_NAME | grep -v $TASKS_DIR | while read child</pre>

This will allow have spaces in descendant names, but, in this case I don't know how to read user input, as +read+ will +read+ next line of the piped output.

Basically, to have more reliable and stable solution I guess it's better to rewrite it in Perl|Python|Ruby|Tcl|JavaScript|PHP|C, but right now I have no time :)) and to be honest don't see practical value for this. So this is just a proof of concept, that works ;))
